### PR TITLE
[image_picker] fix camera on Android 11

### PR DESCRIPTION
**Pull request by Bubu**
## Description

resolveActivity(intent) might return false on android 11+ even though startActivity(intent) would succeed.
Instead of introducing a <queries> entry we can just try to start the
activity and catch the exception if that doesn't work.

See here for a longer form explanation: https://cketti.de/2020/09/03/avoid-intent-resolveactivity/

## Related Issues
Fixes https://github.com/flutter/flutter/issues/85599
flutter/flutter#62669

Some more info on the android 11 behaviour change: https://github.com/flutter/flutter/issues/63727#issuecomment-690070820

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by Bubu** [comment]
@googlebot I signed it!

**Comment by Bubu** [comment]
Hi, is there anything else I can do here?

**Comment by zigapovhe** [comment]
@cyanglaz Is this ready to merge?
@Bubu can you fix conflict with the CHANGELOG.md?

**Comment by Bubu** [comment]
Rebased on master and fixed conflicts.

**Comment by Bubu** [comment]
> This branch has conflicts that must be resolved

*sigh* Any chance that someone might look at this? :thinking:

**Comment by Bubu** [comment]
Any chance to get this looked at?

**Comment by stuartmorgan** [comment]
Thanks for the contribution! We apologize for the long delay in triaging this PR. We’re in the process of overhauling our PR triage system to respond much more quickly, as well as working through the backlog.

@mvanbeusekom Could you add this review to your list?

File: packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java
Comment (renefloor) [inline packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java:364]: We used to finish before creating a temporary file. Now we have created an empty file, I think it would be good to remove the file again before finishing.

**Review by renefloor — CHANGES_REQUESTED** [review]
Looks like a good change to me, especially because Google also recommends `catch (e: ActivityNotFoundException)` in their [blog post](
https://medium.com/androiddevelopers/package-visibility-in-android-11-cc857f221cd9).

However, I think it would be better to clean the empty file.

File: packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java
Comment (Bubu) [reply packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java:375]: Makes total sense, added.

**Review by renefloor — APPROVED** [review]
LGTM

**Comment by renefloor** [comment]
@Bubu can you bring your branch up to date with master once more and update the changelog/pubspec? I'm afraid 0.8.1+1 
is already taken :)

I think there is also 1 broken test (maybe that is new since you fixed the tests).
```
/tmp/cirrus-ci-build/packages/image_picker/image_picker/example/android/app/src/test/java/io/flutter/plugins/imagepicker/ImagePickerDelegateTest.java:197: error: cannot find symbol
    when(mockIntentResolver.resolveActivity(any(Intent.class))).thenReturn(true);
```

**Comment by Bubu** [comment]
> I think there is also 1 broken test (maybe that is new since you fixed the tests).

Ugh, there is... It's been so long since I wrote the changes to the tests here... :see_no_evil: . Let me see if I can make it work again :D.

**Comment by Bubu** [comment]
@renefloor I can't figure out how to run the tests successfully locally (any pointers for that?). I still pushed a commit that should fix it and also (hopefully) resolved the conflicts.  :tada:

**Comment by renefloor** [comment]
@Bubu Currently `takeImageWithCamera_WhenHasCameraPermission_AndNoActivityToHandleCameraIntent_FinishesWithNoCamerasAvailableError` in `ImagePickerDelegateTest` is failing with `java.io.IOException: No such file or directory`.

That's because the test doesn't expect a file to be created (because the camera is not available), so the file system was not mocked.

I added this after line 183 in that test:
```java
    MockedStatic<File> mockStaticFile = Mockito.mockStatic(File.class);
    mockStaticFile
            .when(() -> File.createTempFile(any(), any(), any()))
            .thenReturn(new File("/tmpfile"));
```
and I added the following line at the end of the test:
```java
    mockStaticFile.close();
```

I cannot really comment this as a suggestion on your PR as these are not lines that changed, but I hope it makes sense. The complete test becomes:
```java
  @Test
  public void
      takeImageWithCamera_WhenHasCameraPermission_AndNoActivityToHandleCameraIntent_FinishesWithNoCamerasAvailableError() {
    when(mockPermissionManager.isPermissionGranted(Manifest.permission.CAMERA)).thenReturn(true);

    MockedStatic<File> mockStaticFile = Mockito.mockStatic(File.class);
    mockStaticFile
            .when(() -> File.createTempFile(any(), any(), any()))
            .thenReturn(new File("/tmpfile"));

    doThrow(ActivityNotFoundException.class)
        .when(mockActivity)
        .startActivityForResult(any(Intent.class), anyInt());
    ImagePickerDelegate delegate = createDelegate();
    delegate.takeImageWithCamera(mockMethodCall, mockResult);

    verify(mockResult)
        .error("no_available_camera", "No cameras available for taking pictures.", null);
    verifyNoMoreInteractions(mockResult);
    mockStaticFile.close();
  }
```

It would me nice to also add `mockStaticFile.close();` in `takeImageWithCamera_WritesImageToCacheDirectory`.

**Comment by Bubu** [comment]
:crossed_fingers: 

(I'm flying blind here, when I try to run the tests I get a huge amount of unrelated build errors instead. Let's just hope it works this time :))

**Comment by Bubu** [comment]
We are now failing at a different test. I don't quite understand why my MR would be changing the behavior for that test. `takeImageWithCamera_WhenHasCameraPermission_AndAnActivityCanHandleCameraIntent_LaunchesTakeWithCameraIntent` should have been creating files even before this change?

Did some merge go wrong here?

**Comment by renefloor** [comment]
I'm at my phone at the moment, but make sure to open image_picker/example/android and not image_picker/android. You might need to run the flutter example once or maybe you are missing parts of the SDK.

**Comment by renefloor** [comment]
@Bubu The test fails because it depends on another test not closing the mock, so it also doesn't run when you only test that 1 test.

You can fix it by moving the mock to the setup and add a teardown. This is working: https://github.com/Baseflow/flutter-plugins/commit/149194c80f862ca7b8085044cb007e7476013320

**Comment by Bubu** [comment]
@renefloor rebased again, which should take care of the tests

I'm unsure if I need to make it version `+4` or `+5` now? `+4` seems to exist in git but isn't merged into `master`? Should I base the whole MR on the `+4` version?

**Comment by Bubu** [comment]
There's a Ci build failure for `:video_player` which seems unrelated to this MR>

**Comment by renefloor** [comment]
I think this seems fine.

>Could not GET 'https://google.bintray.com/exoplayer/com/google/android/exoplayer/exoplayer-hls/2.12.1/exoplayer-hls-2.12.1.pom'. Received status code 403 from server: Forbidden

I think this should resolve itself in time. I'll try to rerun this tomorrow.

**Comment by stuartmorgan** [comment]
> I think this seems fine.
> 
> > Could not GET 'https://google.bintray.com/exoplayer/com/google/android/exoplayer/exoplayer-hls/2.12.1/exoplayer-hls-2.12.1.pom'. Received status code 403 from server: Forbidden
> 
> I think this should resolve itself in time. I'll try to rerun this tomorrow.

This is fixed, but requires merging in the latest master (we had to remove the dependency on Bintray).

**Comment by Bubu** [comment]
Rebased. Let's see how it goes :).

**Comment by stuartmorgan** [comment]
@blasten Do you want to do a final review on this?

**Comment by Bubu** [comment]
Hey everyone, any progress here? :-)

File: packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java
Comment (blasten) [inline packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java:374]: this may throw `SecurityException`, so it must be catch, so it doesn't end in a bad state. 
https://docs.oracle.com/javase/7/docs/api/java/io/File.html#delete()

**Review by blasten — CHANGES_REQUESTED** [review]
requires to catch the exception

File: packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java
Comment (Bubu) [reply packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java]: Done, I think. (Not sure what to do about that potential exception really but to ignore it.)

File: packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java
Comment (blasten) [inline packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java:295]: how do we know that `startActivityForResult` deletes the file, or we don't know? Is this specified somewhere?

Comment (Bubu) [reply packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java:295]: I'm unsure what you are referring to here. The case where `startActivityForResult` succeeds isn't modified in this PR and should be working as before. If it doesn't succeed the tmpfile we just created before trying the call is deleted again. (which theoretically might fail, but if the system doesn't let us delete a file we just created there's nothing we can do about that. :shrug: )

Comment (blasten) [reply packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java:295]: I see.  I wasn't sure about the lifecycle of the file

**Review by blasten — APPROVED** [review]

**Comment by ManuLpz4** [comment]
Please ♥️

**Comment by stuartmorgan** [comment]
@Bubu Could you resolve the conflicts so we move forward with landing this?

**Comment by Bubu** [comment]
@stuartmorgan hopefully all done correctly.

File: packages/image_picker/image_picker/CHANGELOG.md
Comment (stuartmorgan) [inline packages/image_picker/image_picker/CHANGELOG.md:1]: You need to update the version here to match what's being published.

Comment (Bubu) [reply packages/image_picker/image_picker/CHANGELOG.md]: Done.

**Comment by Bubu** [comment]
There's now conflicts again here now... :thinking:

**Comment by stuartmorgan** [comment]
That's my mistake for not following up after merging in master to fix the infra issues.

**Pull request merged**
